systemtests/nidigital_codemodules.py

```diff
@@ -0,0 +1,98 @@
+        )
+        session.apply_levels_and_timing("nidigital", "nidigital")
+        for pattern_file_path in tsm_context.nidigital_project_pattern_file_paths:
+            session.load_pattern(pattern_file_path)
+        tsm_context.set_nidigital_session(instrument_name, session)
+
+
+@nitsm.codemoduleapi.code_module
+def measure_ppmu(
+    tsm_context: SemiconductorModuleContext,
+    pins,
+    expected_instrument_names,
+    expected_pin_set_strings,
+):
+    pin_query, sessions, pin_set_strings = tsm_context.pins_to_nidigital_sessions_for_ppmu(pins)
+    expected_instrument_pin_sets = set(zip(expected_instrument_names, expected_pin_set_strings))
+    valid_pin_sets = []
+
+    for session, pin_set_string in zip(sessions, pin_set_strings):
+        # call some methods on the session to ensure no errors
+        session.pins[pin_set_string].ppmu_aperture_time = 4e-6
+        session.pins[
+            pin_set_string
+        ].ppmu_aperture_time_units = nidigital.PPMUApertureTimeUnits.SECONDS
+        session.pins[pin_set_string].ppmu_output_function = nidigital.PPMUOutputFunction.CURRENT
+        session.pins[pin_set_string].ppmu_current_level_range = 2e-6
+        session.pins[pin_set_string].ppmu_current_level = 2e-6
+        session.pins[pin_set_string].ppmu_voltage_limit_high = 3.3
+        session.pins[pin_set_string].ppmu_voltage_limit_low = 0
+        session.pins[pin_set_string].ppmu_source()
+        session.pins[pin_set_string].ppmu_measure(nidigital.PPMUMeasurementType.CURRENT)
+        session.abort()
+
+        # check instrument pin set we received is in the set of instrument pin sets we expected
+        actual_instrument_pin_set = (session.io_resource_descriptor, pin_set_string)
+        num_pins_for_session = len(pin_set_string.split(","))
+        valid_pin_sets.extend(
+            [actual_instrument_pin_set in expected_instrument_pin_sets] * num_pins_for_session
+        )
+        expected_instrument_pin_sets -= {actual_instrument_pin_set}
+
+    pin_query.publish(valid_pin_sets, "ValidPinSetStrings")
+    num_missing_pin_sets = [len(expected_instrument_pin_sets)] * len(valid_pin_sets)
+    pin_query.publish(num_missing_pin_sets, "NumMissingPinSetStrings")
+
+
+@nitsm.codemoduleapi.code_module
+def measure_pattern(
+    tsm_context: SemiconductorModuleContext, pins, expected_instrument_names, expected_site_lists
+):
+    pin_query, sessions, site_lists = tsm_context.pins_to_nidigital_sessions_for_pattern(pins)
+    expected_instrument_site_lists = set(zip(expected_instrument_names, expected_site_lists))
+    valid_site_lists = []
+    re_pattern = re.compile(r"\s*site(\d+)")
+
+    for session, site_list in zip(sessions, site_lists):
+        # call some methods on the session to ensure no errors
+        session.pins[pins].configure_active_load_levels(0.0015, -0.024, 2.0)
+        session.pins[pins].configure_voltage_levels(0.1, 3.3, 0.5, 2.5, 5.5)
+        session.commit()
```

Comment:
I would omit these lines if possible. They are technically incorrect because they are configuring pins for all sites even though the code should only be affecting the sites in site_list.